File: 16. 最接近的三数之和.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number}
 */
var threeSumClosest = function(nums, target) {
    let out = 0;
    let minDist = Infinity;

    nums.sort((a, b)=>a - b);

    for (let i = 0; i < nums.length; i++) {
        if (i > 0 && nums[i] === nums[i - 1]) continue;

        let left = i + 1;
        let right=  nums.length - 1;
        while (left < right) {
            const sum = nums[i] + nums[left] + nums[right];
            const dist = Math.abs(sum - target);

            if (sum === target) return target;
            if (dist < minDist) {
                minDist = dist;
                out = sum;
            }
            if (sum > target) {
                right--;
            } else {
                left++;
            }
        }
    }
    return out;
};

File: 16. 最接近的三数之和.cpp
class Solution {
public:
    int threeSumClosest(vector<int>& nums, int target) {
        sort(nums.begin(), nums.end());

        int out = nums[0] + nums[1] + nums[2];
        for (int i = 0; i < nums.size(); i++) {
            int left = i + 1;
            int right = nums.size() - 1;
            while (left < right) {
                int sum = nums[left] + nums[right] + nums[i];
                if (abs(sum - target) < abs(out - target)) out = sum;
                if (sum < target) {
                    left++;
                } else {
                    right--;
                }
            }
        }

        return out;
    }
};